### PR TITLE
Communities: Share airdrop address in non token-gated community

### LINKED ISSUE #18081: Communities: Share airdrop address in non token-gated community
# Feature Issue

### User Story

- As a User
  - Joining a community
- I want to select my airdrop address from my full account / address list
- So that I can maintain access to the community if it becomes token-gated
- And I can be granted community permissions

### Technical tasks
- [x] Use component `settings.settings-item.view/view` to display a summary of the selected airdrop addresses.
- [x] Use component `quo.components.list-items.account.view` to display the list of possible airdrop addresses.
  - PR with the above 2 changes: https://github.com/status-im/status-mobile/pull/18189
- [ ] Answer questions in the `Questions` section below.
- [x] Add validations in status-go to fulfill the requirement that airdrop addresses should be a subset of the selected addresses to share.
      
**Note**: It is okay to focus on the functionality first and polish in future PRs.

### Questions

- Should airdrop addresses be a subset of the previous address selection?
  - Yes, see this [Figma node](https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=15331-293400&mode=design&t=UNMQM7sTa7XQyn15-0 ) from designers 
- What is the default airdrop account? First account in list?
- What should happen when deselecting the *selected airdrop account* from the list of *shared accounts*?
  - Does the airdrop address change to the next address in the selected list?
- Does the UI / backend prevent deselection of a selected airdrop account?
- What is the behaviour when the user selects an account derived from an imported private key, then joins the community with that private key account, then recovers their *Status Account* and doesn’t still have access to the private key account when the user attempts to access the community?

### PR DESCRIPTION
**Pull request by FFFra**
fixes #18081 

### Summary

This pull request addresses the first two subtasks of #18081, encompassing:

- Displaying a summary of the selected airdrop addresses.
- Presenting the list of potential airdrop addresses.

[Figma](https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=15331-286958&mode=design&t=esUiSc1hu4Tc066w-0)

Please check changes here:

Before refactor, using bottom-sheet:

https://github.com/status-im/status-mobile/assets/5999878/b47e4cf8-1304-4d82-8996-ab3fcba8f46d

After the refactor, using a new screen:

https://github.com/status-im/status-mobile/assets/5999878/d4d487e1-8386-4071-8ee7-9355e4e93c71

### Review notes
For now, we are not been pixel perfect. To actually select the address is the next scope.

This feature is currently under a flag that is not enabled in the develop branch, so these changes do not require manual QA.

##### Platforms
- Android
- iOS

##### Functional

- public chats
- group chats
- wallet / transactions
- networks

### Steps to test
- Navigate to `src/status_im2/config.cljs` and set `community-accounts-selection-enabled?` to `true`.
- Ensure you have some wallets. If not, you can create them in the wallet tab.
- Create a community on desktop and invite another user.
- When the invited user attempts to join this community, the option to select airdrop addresses should appear.

status: ready

### CODE REVIEW

File: src/status_im2/contexts/communities/actions/accounts_selection/view.cljs [hunk 1]
```diff
@@ -70,12 +72,20 @@
...
+                                                              {:accounts       accounts
+                                                               :logo-uri       (get-in images
+                                                                                       [:thumbnail :uri])
+                                                               :community-name name}])}])
                       :description       :text
                       :action            :arrow
                       :label             :preview
                       :label-props       {:type :accounts
-                                          :data (take 1 accounts)}
+                                          ;; TODO: Substitute the provided data with the selected airdrop
```
Comment (J-Son89) [inline src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: please create an issue and link in the TODO.  👍

Comment (FFFra) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: We have the issue already, gonna link it. Good idea!

Comment (ajayesivan) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: I recall there was a discussion about discontinuing the use of TODO comments and relying solely on GitHub issues to track tasks.

File: src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    :emoji         (:emoji item)}])
+
+(defn- accounts-list
+  [{:keys [accounts]}]
+  [rn/view {:style style/account-list-container}
+   (when (seq accounts)
+     [rn/flat-list
+      {:data      accounts
+       :render-fn render-item
+       :key-fn    :id}])])
```
Comment (ajayesivan) [inline src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs]: I don't remember seeing an `:id` in the `accounts` map. Could you please double check?

File: src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+(defn- airdrop-addresses
+  [{:keys [accounts logo-uri community-name]}]
```
Comment (ajayesivan) [inline src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs]: Given that we already have the accounts data in our app db, I suggest directly subscribing to it here instead of passing it down as props. I'm unsure about the preferred approach for handling this. @J-Son89, what do you think? 

cc: @ilmotta

File: src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      :size       32
+      :icon-only? true
+      :on-press   #(not-implemented/alert)} 
+     :i/info]] 
+   [quo/context-tag
+    {:type            :community
+     :size            24
+     :community-logo  logo-uri
+     :community-name  community-name
+     :container-style {:margin-top 8}}]
```
Comment (ajayesivan) [inline src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs]: As you mentioned [here](https://github.com/status-im/status-mobile/pull/18180/files#r1427104905), we can use `drawer-top` component.

File: src/status_im2/contexts/communities/actions/accounts_selection/view.cljs (hunk 1 above)
Comment (J-Son89) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: Yeah it's preferable to do that 👌

**Review by ajayesivan — COMMENTED** [review]
There is a delay when opening the airdrop sheet, check out the video. I'm implementing it as a screen and the performance is better. 

https://github.com/status-im/status-mobile/assets/19279756/a2696e9b-c50c-40de-bc42-57c7377b8078

File: src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs (hunk 3 above)
Comment (cammellos) [reply src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs]: In this case, the subscription should be here, @ajayesivan is correct.
If I follow the code correctly, in the `on-press` event we initialise view:

```
                      :on-press          #(rf/dispatch
                                           [:show-bottom-sheet
                                            {:content (fn [] [airdrop-addresses/view
                                                              {:accounts       accounts
                                                               :logo-uri       (get-in images
                                                                                       [:thumbnail :uri])
                                                               :community-name name}])}])
```

`accounts` is a subscription, so if anything changes in `accounts` the `on-press` event will change. 

The problem is that if you already clicked, and the view is displayed already and `accounts` in the db changes, this change won't be reflect in the UI, unless you close and re-open the sheet, making it effectively non-reactive.

So we should subscribe to accounts in this component, and no need to pass accounts on the on-press event. (things might change once we have a selection etc, but for now, that would be the correct implementation I believe).

File: src/status_im2/contexts/communities/actions/accounts_selection/view.cljs [hunk 5]
```diff
@@ -41,6 +42,7 @@
                                          :accounts
                                          vals
                                          (map #(assoc % :customization-color (:color %))))]
+
```
Comment (cammellos) [inline src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: @FFFra I know you haven't written this code, but this subscription is way to broad, it's subscribing to the whole wallet and pulling just accounts out.
This means that anything that changes in `wallet` (regardless of whether `:accounts` is impacted), it will re-render the whole thing.

All of this should be in a subscription:

```
        accounts                    (->> (rf/sub [:wallet])
                                         :accounts
                                         vals
                                         (map #(assoc % :customization-color (:color %))))]
```
it should look something like:

`accounts-with-customization-color (rf/sub [:wallet/accounts-with-customization-color])` (or whatever naming, the idea is that all the work should be in a subscription and that should be at least a layer-2 subscription, data manipulation in the view should be kept to a minimum.

As a side node, this is a bit an issue with having nested structs in the database, they can help and they are neater, but it's more prone to error, so we should always be careful when we subscribe to make sure we only subscribe to what is strictly needed for the view and no more.

File: src/status_im2/contexts/communities/actions/accounts_selection/view.cljs [hunk 6]
```diff
@@ -70,12 +72,20 @@
                                           :data accounts}
                       :description-props {:text (i18n/label :t/all-addresses)}}
                      {:title             (i18n/label :t/for-airdrops)
-                      :on-press          not-implemented/alert
+                      :on-press          #(rf/dispatch
+                                           [:show-bottom-sheet
+                                            {:content (fn [] [airdrop-addresses/view
```
Comment (cammellos) [inline src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: Because you are passing the data on a `on-press`, this effectively makes everything in that view non-reactive. I think you should just initialise the view and subscribe within that view, so that if things changes while the view is renderered, things change accordingly.

File: src/status_im2/contexts/communities/actions/accounts_selection/view.cljs (hunk 5 above)
Comment (J-Son89) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs:45]: @FFFra - can you explain why we need to add customization-color here?

Comment (J-Son89) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: + 1 on this.

Comment (ajayesivan) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: That's my code 😬.

Comment (ajayesivan) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: I have the same thing in my PR too.

Comment (ajayesivan) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: I can update my PR with a subscription like @cammellos suggested and once that's merged @FFFra can make use of that. Or should we go with a follow-up issue approach?

Comment (cammellos) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: up to you, since it's not going to be used and we can merge things quite quickly, either are fine, best to merge as quick as possible since it's all behind a flag, and follow up, to avoid conflicts, but up to you, as long as we don't forget.

Comment (ajayesivan) [reply src/status_im2/contexts/communities/actions/accounts_selection/view.cljs]: I think followup is the best approach to avoid conflicts. I have created[ this issue ](https://github.com/status-im/status-mobile/issues/18198)and assigned myself, will raise a PR after this one is merged.

@FFFra you can skip this suggestion for now.

File: src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs (hunk 2 above)
Comment (FFFra) [reply src/status_im2/contexts/communities/actions/airdrop_addresses/view.cljs]: Done! Thanks!

**Comment by FFFra** [comment]
Hello, team/ I've updated the PR description with the new video so we can compare :)

**Review by ajayesivan — APPROVED** [review]
Nice work!

File: src/quo/core_spec.cljs
Comment (ajayesivan) [inline src/quo/core_spec.cljs:1]: These changes are unrelated to this PR. I moved disabled tests into a comment block to prevent `make lint-fix` messups.

`make lint-fix` was creating duplicate comments:
Screenshot 2023-12-21 at 10 42 03

**Comment by ajayesivan** [comment]
The changes introduced in this PR are disabled in develop under a flag, so it doesn't require manual qa. I think failed e2e is not related to this PR. @status-im/mobile-qa Could you please check and confirm?

**Comment by pavloburykh** [comment]
@ajayesivan thank you for the PR. Failed e2e are not PR related. Ready for merge.
